### ISSUE TITLE
Pinned memory leak

### ISSUE THREAD

**Issue by nizamov** [cupy/cupy#4775]
Hi, there

seems that I have found some kind of a strange memory allocation bug. I am running Debian testing, which hast cuda 11.2 and cython 29.21. So I had to build cupy ("pip3 install cupy" - no errors reported except missing cudnn/cublas etc).

Now, the problem is that I wanted to benchmark memory transfer between host memory and device memory. In general, this is nothing more than measuring the timing of many transfers of the same array. So memory allocation had to be really negligible, but with cupy it is not. It is growing until it gets killed by OOM-killer. I had the similar test in Numba without any problems. 

Here is a simplified code that triggers such behavior on my PC (quite an old Xeon E3-1220 with 16Gb DDR3 memory, python 3.9.1).  If there is no explicit release of memory buffers, the memory occupied by python grows steadily. Actually it should not. I have added a small time delay to the cycle to be able to observe this behavior and found another strange thing. If time delay is longer than 0.2s (I suspect that the value depends on the hardware), then the memory occupation gets capped at 3.4 Gb and test passes. For comparison, with pinned_mempool.free_all_blocks() memory occupation remains at the level around 2.4Gb. 

It seems to me, that there is some automatic garbage collection which can not cope with a rate that this test allocates/deallocates pinned memory. But works to at least some extent, if the rate is decreased by addition of sufficient sleeping time. Since I do not observe such problem with Numba on the same configuration, I guess that this bug is not in linux/python/cuda stack but rather in cupy itself.  

```
import numpy as np
import cupy as cp
import time

N = 10*1024*1024

cpu_data = np.zeros(N, dtype=np.float32)

mempool = cp.get_default_memory_pool()
pinned_mempool = cp.get_default_pinned_memory_pool()

for i in range(5000):
    gpu_data = cp.asarray(cpu_data)

    if i % 100 == 0:
        print(i, mempool.used_bytes(), mempool.total_bytes(), pinned_mempool.n_free_blocks())    

        time.sleep(0.1)
        #pinned_mempool.free_all_blocks()

print("Ok")

```

Message in dmesg when test gets killed:

```
Out of memory: Killed process 7203 (python3) total-vm:17809796kB, anon-rss:115924kB, file-rss:75444kB, shmem-rss:11151360kB, pgtables:22756kB oom_score_adj:0
oom_reaper: reaped process 7203 (python3), now anon-rss:0kB, file-rss:74732kB, shmem-rss:11151360kB

```

Output of `python -c 'import cupy; cupy.show_config()'`

```
OS                           : Linux-5.10.0-3-amd64-x86_64-with-glibc2.31
CuPy Version                 : 8.5.0
NumPy Version                : 1.19.5
SciPy Version                : 1.6.0
Cython Build Version         : 0.29.22
CUDA Root                    : /usr
CUDA Build Version           : 11020
CUDA Driver Version          : 11020
CUDA Runtime Version         : 11020
cuBLAS Version               : 11301
cuFFT Version                : 10400
cuRAND Version               : 10203
cuSOLVER Version             : (11, 0, 2)
cuSPARSE Version             : 11301
NVRTC Version                : (11, 2)
Thrust Version               : 101100
CUB Build Version            : 101100
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
Device 0 Name                : GeForce GTX 1660 SUPER
Device 0 Compute Capability  : 75
```

**Comment by nizamov** [comment]
I guess, that the reason is asynchronous memory transfer. Seems that CuPy allocates pinned memory, necessary for the transfer, and keeps it until the transfer over PCIe finishes. I see that memory occupation grows more than 3GB only if more than 7-8Gbytes/second bandwidth is required by test script. This is around the bandwidth that I get with Numba and roughly corresponds to the expected host-gpu memory bandwidth with PCIe-3x16 and DDR3.  

Adding explicit synchronisation, like shown here https://github.com/cupy/cupy/issues/1317 seems to be solving the issue of overcommiting the host memory. BTW, Numba is synchronous (if not using streams explicitly). 

cupy.cuda.stream.get_current_stream().synchronize()

**Comment by kmaehashi** [comment]
Thank you for the report! This is interesting, we have a guard for the case that pinned memory cannot be allocated for the transfer; it should fall back to sync mode if pinned memory cannot be allocated.
https://github.com/cupy/cupy/blob/2fd13b80ba108bcc12d102fd7fa9d5da0ced7eaf/cupy/core/core.pyx#L2375-L2388
Could you paste the `$ ulimit -a` output? If `locked-in-memory size`  is restricted, could you try your code again after setting it to `unlimited`?

**Comment by nizamov** [comment]
This is what "ulimit -a" shows on my PC running Debian 11 (testing at the moment). On two other PCs, running Ubuntu 20.04 and Debian 10 (stable), output is pretty much the same, excluding max locked memory - they have 65536. 

```
real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) 0
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 0
file size                   (blocks, -f) unlimited
pending signals                     (-i) 63628
max locked memory           (kbytes, -l) 2044941
max memory size             (kbytes, -m) unlimited
open files                          (-n) 1024
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 0
stack size                  (kbytes, -s) 8192
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) 63628
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited

```

After changing this limit to unlimited (and having checked that this was applied) I run the test script with the same result - memory allocation steadily grows until PC hangs. In my first post I said that OOM-killer was killing the process, but often it was also just a hard freeze. 

Now I am puzzled - seems that ulimits had to prevent CuPy from pinning more than 2Gb of memory. And that would be safe. But it did not happen, something is really going wrong. In man setrlimit I see the following statement:

```Since Linux 2.6.9, no limits are placed on the amount of memory that a privileged process may lock, and this limit instead governs the amount of memory that an unprivileged process may lock.```

Could it be that CuPy allocates memory through some library with elevated privileges and thus is not subject to ulimit? I am running with an ordinary user (it belongs to "video" group though), not a root.

**Comment by agnikit** [comment]
I face exactly the same issue. My code includes a lot of cpu/gpu transfers and the host memory consumption is huge, to the point that the system crashes. Also, as already noted, ulimit doesn't work as expected. Is there currently any workaround for this issue? Manually freeing pinned memory is the way to go?

**Comment by kmaehashi** [comment]
I'll dig into this one in more detail, but disabling the pinned memory pool should workaround the problem in your cases.

https://docs.cupy.dev/en/latest/user_guide/memory.html#changing-memory-pool

```py
cupy.cuda.set_pinned_memory_allocator(None)
```

**Comment by wohali** [comment]
Another data point: even after calling `cupy.cuda.set_pinned_memory_allocator(None)` I am still getting the PerformanceWarning console errors. Even after trying the `ulimit` change mentioned above, I am not getting machine hangs, but I am still getting the warnings:

```python
/home/user/.cache/pypoetry/virtualenvs/app-d-p5eUlI-py3.11/lib/python3.11/site-packages/cupy/_creation/from_data.py:88: PerformanceWarning: Using synchronous transfer as pinned memory (134384192 bytes) could not be allocated. This generally occurs because of insufficient host memory. The original error was: cudaErrorMemoryAllocation: out of memory
  return _core.array(a, dtype, False, order, blocking=blocking)
/home/user/.cache/pypoetry/virtualenvs/app-d-p5eUlI-py3.11/lib/python3.11/site-packages/cupy/_creation/from_data.py:88: PerformanceWarning: Using synchronous transfer as pinned memory (133589652 bytes) could not be allocated. This generally occurs because of insufficient host memory. The original error was: cudaErrorMemoryAllocation: out of memory
  return _core.array(a, dtype, False, order, blocking=blocking)
```

Of course, as soon as this happens, performance tanks.

One confounding factor may be that this is under WSL2. Per Performance Monitor, shared GPU memory usage never seems to exceed 4GB, though plenty more is available.

System details: RTX 4090 (24GB), 64GB CPU memory (never exceeding 50% utilisation), CUDA 12.8.1, `cupy-cuda12x==13.4.1`. 

Is there any way to e.g. catch these warnings and attempt to free all pinned blocks? Or are there other options possible?

**Comment by kmaehashi** [comment]
> One confounding factor may be that this is under WSL2. Per Performance Monitor, shared GPU memory usage never seems to exceed 4GB, though plenty more is available.

The "pinned memory" is an allocation on host memory, not GPU device memory. Pinned memory is used as a staging buffer to copy data from host to GPU memory.

> System details: RTX 4090 (24GB), 64GB CPU memory (never exceeding 50% utilisation), CUDA 12.8.1, cupy-cuda12x==13.4.1.

Check output of the `free` command on your WSL2 terminal. It may worth try increasing the total memory allocated to your WSL virtual machine.

**Comment by wohali** [comment]
@kmaehashi Perhaps, but with `cupy.cuda.set_pinned_memory_allocator(None)` shouldn't the warnings be completely suppressed? They're not.

**Comment by kmaehashi** [comment]
> with `cupy.cuda.set_pinned_memory_allocator(None)` shouldn't the warnings be completely suppressed?

No, this just disables pooling pinned memory allocation for future reuse. Allocation of pinned memory is anyway required to transfer the data.

**Comment by kmaehashi** [comment]
I took some time to investigate the original issue but couldn't reproduce locally.
